src/resampler.rs

```diff
@@ -0,0 +1,279 @@
+{
+}
+impl DataType for f32 {}
+impl DataType for f64 {}
+
+/// The Sample trait represents a single sample in a time series.
+pub trait Sample: Clone + Debug + Copy + Default {
+    type Value;
+    fn new(timestamp: DateTime<Utc>, value: Option<Self::Value>) -> Self;
+    fn timestamp(&self) -> DateTime<Utc>;
+    fn value(&self) -> Option<Self::Value>;
+}
+
+/// The ResamplingFunction enum represents the different resampling functions
+/// that can be used to resample a channel.
+pub enum ResamplingFunction<T: DataType, S: Sample<Value = T>> {
+    /// Calculates the average of all samples in the time step (ignoring None
+    /// values)
+    Average,
+    /// Calculates the sum of all samples in the time step (ignoring None
+    /// values)
+    Sum,
+    /// Calculates the maximum value of all samples in the time step (ignoring
+    /// None values)
+    Max,
+    /// Calculates the minimum value of all samples in the time step (ignoring
+    /// None values)
+    Min,
+    /// Uses the last sample in the time step. If the last sample is None, the
+    /// resampling function will return None.
+    Last,
+    /// Counts the number of samples in the time step (ignoring None values)
+    Count,
+    /// A custom resampling function that takes a closure that takes a slice of
+    /// samples and returns an optional value.
+    Custom(CustomResamplingFunction<S, T>),
+}
+
+impl<T: DataType, S: Sample<Value = T>> ResamplingFunction<T, S> {
+    pub fn apply(&mut self, samples: &[&S]) -> Option<T> {
+        match self {
+            Self::Average => Self::Sum
+                .apply(samples)
+                .and_then(|sum| Self::Count.apply(samples).map(|count| sum.div(count))),
+            Self::Sum => samples.iter().filter_map(|s| s.value()).sum1(),
+            Self::Max => samples.iter().filter_map(|s| s.value()).max_by(|a, b| {
+                a.partial_cmp(b).unwrap_or_else(|| {
+                    if a.is_finite() {
+                        std::cmp::Ordering::Greater
+                    } else {
+                        std::cmp::Ordering::Less
+                    }
+                })
+            }),
+            Self::Min => samples.iter().filter_map(|s| s.value()).min_by(|a, b| {
+                a.partial_cmp(b).unwrap_or_else(|| {
+                    if a.is_finite() {
+                        std::cmp::Ordering::Less
+                    } else {
+                        std::cmp::Ordering::Greater
+                    }
+                })
+            }),
+            Self::Last => samples.last().and_then(|s| s.value()),
+            Self::Count => Some(
+                T::from_usize(samples.iter().filter_map(|s| s.value()).count())
+                    .unwrap_or_else(|| T::zero()),
+            ),
+            Self::Custom(f) => f.as_mut()(samples),
+        }
+    }
+}
+
+impl<T: DataType, S: Sample<Value = T>> Default for ResamplingFunction<T, S> {
+    fn default() -> Self {
+        Self::Average
+    }
+}
```

Comment:
I think this can be auto derived using `#[derive(Default)]` by using the `#[default]` attribute right above the variant `Average`.